src/server_manager/web_app/ui_components/outline-per-key-data-limit-dialog.ts

```diff
@@ -0,0 +1,380 @@
+          auto-validate
+          value=${this._initialValue()}
+          size="7"
+          @keyup=${this._setSaveButtonDisabledState}
+        >
+        </paper-input>
+        <paper-dropdown-menu id="unitsDropdown" noink>
+          <paper-listbox
+            id="unitsListbox"
+            slot="dropdown-content"
+            attr-for-selected="name"
+            selected="${this._initialUnit()}"
+          >
+            <paper-item name="GB">${this._getInternationalizedUnit(1000000000)}</paper-item>
+            <paper-item name="MB">${this._getInternationalizedUnit(1000000)}</paper-item>
+          </paper-listbox>
+        </paper-dropdown-menu>
+      </div>
+    `;
+  }
+
+  private _queryAs<T extends HTMLElement>(selector: string): T {
+    return this.shadowRoot.querySelector(selector) as T;
+  }
+
+  private get _input(): PaperInputElement {
+    return this._queryAs<PaperInputElement>('#dataLimitInput');
+  }
+
+  private _dataLimitValue() {
+    return Number(this._input?.value) ?? 0;
+  }
+
+  private _dataLimitType() {
+    return this._queryAs<PaperListboxElement>('#unitsListbox').selected as 'GB' | 'MB';
+  }
+
+  private _getInternationalizedUnit(bytes: number) {
+    return formatBytesParts(bytes, this._language).unit;
+  }
+
+  private _initialUnit(): 'GB'|'MB' {
+    return bytesToDisplayDataAmount(this._activeDataLimit())?.unit || 'GB';
+  }
+
+  private _initialValue() {
+    return bytesToDisplayDataAmount(this._activeDataLimit())?.value.toString() || '';
+  }
+
+  private _activeDataLimit(): number|undefined {
+    // Returns the limit which currently is enforced on this key, or undefined if there is none.
+    return this._keyDataLimitBytes ?? this._serverDefaultLimitBytes;
+  }
+
+  private async _setCustomLimitTapped() {
+    this._showMenu = !this._showMenu;
+    if (this._showMenu) {
+      await this.updateComplete;
+      this._input?.focus();
+    }
+  }
+
+  private _setSaveButtonDisabledState() {
+    this._enableSave = !this._input?.invalid ?? false;
+  }
+
+  private _sendSaveEvent() {
+    const change = this._dataLimitChange();
+    const eventName = `${change === Change.REMOVED ? 'Remove' : 'Save'}PerKeyDataLimitRequested`;
+
+    this.dispatchEvent(new CustomEvent(eventName, {
+      // The target of this event is appRoot, not the dialog.  We send the full dialog element so
+      // that the app can control the dialog accordingly to user input.
+      detail: {dialog: this},
+      // Required for the event to bubble past a shadow DOM boundary
+      bubbles: true,
+      composed: true,
+    }));
+  }
+
+  /**
+   * Calculates what type of change, or none at all, the user made.
+   */
+  private _dataLimitChange(): Change {
+    if (this._showMenu) {
+      if (this._keyDataLimitBytes === undefined) {
+        // The user must have clicked the checkbox and input a limit.
+        return Change.SET;
+      }
+      const inputLimit = displayDataAmountToBytes(this.inputDataLimit());
```

Comment:
displayDataAmountToBytes doesn't belong in the UI. Move it to the app code instead.

Reply:
I find it cleaner to deal with `number` as a vocabulary type in between components instead of passing around `DisplayDataAmount`s.  This way the app doesn't need to know or care how the dialog is formatting the data limit, and the UI and the server don't need to know what the other one is doing. Using bytes in `DisplayAccessKey` as well instead of `DisplayDataAmount` helped us in d5dd4af as well.

  All of the byte <=> display code can live in the UI and then we don't need to worry about particular roundings or display logic in event handlers, `refreshTransferStats`, `convertToUiAccessKey`, etc.